### PR TITLE
feat(backend): add partners portal users multi factor authentication

### PR DESCRIPTION
**Pull request by pbn4**
# Pull Request Template

## Issue Overview

This PR addresses #2244 

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This change adds multifactor authentication to user_accounts for which column `mfa_enabled` is true. 

Endpoints:

1. `POST /auth/request-mfa-code` -> allows to retrieve MFA code by sending user credentials (email and password). 
2. `POST /auth/mfa-info` -> allows to retrieve MFA info for an account by sending user credentials (email and password). Response:
```
export class GetMfaInfoResponseDto {
  maskedPhoneNumber?: string
  email?: string
  isMfaEnabled: boolean
  mfaUsedInThePast: boolean
}
```
Frontend should call this first to properly build MFA option form:
- if isMfaEnabled is false then user can log in with just email and password,
- if `maskedPhoneNumber` is `undefined` and `mfaUsedInThePast` is false then user can provide phone number for the first time in request-mfa-code endpoint

## How Can This Be Tested/Reviewed?

Provide instructions so we can review.

Describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [x] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [ ] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

**Comment by emilyjablonski** [comment]
What do you think is the best way to test this PR?

**Comment by pbn4** [comment]
> What do you think is the best way to test this PR?

Good question. It's pure API change so it's doable e.g. by postman or an autogenerated JS client. Try following e2e test scenario [here](https://github.com/bloom-housing/bloom/pull/2291/files#diff-bbfd947f74508795ca57e30a4d194d5053a08828ee3fd38d4b4ce2068b54228fR794) which basically retrieves MFA information for a given user and then provides MFA credentials. The important part is that your user must have `mfa_enabled` equal to true in `user_accounts` table in the DB. 

Testing an SMS requires setting up few environment variables when running backend (cc: @seanmalbert, could you provide those, please?): TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN, TWILIO_PHONE_NUMBER.

**Comment by seanmalbert** [comment]
> Testing an SMS requires setting up few environment variables when running backend (cc: @seanmalbert, could you provide those, please?): TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN, TWILIO_PHONE_NUMBER.

I added these to the dev backend settings in heroku.

File: backend/core/src/auth/dto/request-mfa-code.dto.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  @IsString({ groups: [ValidationsGroupsEnum.default] })
+  password: string
+
+  @Expose()
+  @IsEnum(MfaType, { groups: [ValidationsGroupsEnum.default] })
+  mfaType: MfaType
+
+  @Expose()
+  @IsOptional({ groups: [ValidationsGroupsEnum.default] })
+  @IsPhoneNumber(null, { groups: [ValidationsGroupsEnum.default] })
```
Comment (YazeedLoonat) [inline backend/core/src/auth/dto/request-mfa-code.dto.ts:23]: This seems to conflict with how twillio handles phone number formats for us domestic numbers. I think we should probably remove it

**Review by YazeedLoonat — CHANGES_REQUESTED** [review]
This looks good, just a few small changes/suggestions

for anyone else that wants to test this, you'll need to work with sean to get your phone number verified for the test phone number to send an sms to you

**Review by YazeedLoonat — APPROVED** [review]
LGTM

**Pull request merged**